commit 45239ba39a5279e9efc671774e2eef29df4d2484
Author: Pawan Gupta <pawan.kumar.gupta@linux.intel.com>
Date:   Fri Oct 25 12:14:58 2024 -0500

    x86/cpu: Add CPU type to struct cpuinfo_topology
    
    Sometimes it is required to take actions based on if a CPU is a performance or
    efficiency core. As an example, intel_pstate driver uses the Intel core-type
    to determine CPU scaling. Also, some CPU vulnerabilities only affect
    a specific CPU type, like RFDS only affects Intel Atom. Hybrid systems that
    have variants P+E, P-only(Core) and E-only(Atom), it is not straightforward to
    identify which variant is affected by a type specific vulnerability.
    
    Such processors do have CPUID field that can uniquely identify them. Like,
    P+E, P-only and E-only enumerates CPUID.1A.CORE_TYPE identification, while P+E
    additionally enumerates CPUID.7.HYBRID. Based on this information, it is
    possible for boot CPU to identify if a system has mixed CPU types.
    
    Add a new field hw_cpu_type to struct cpuinfo_topology that stores the
    hardware specific CPU type. This saves the overhead of IPIs to get the CPU
    type of a different CPU. CPU type is populated early in the boot process,
    before vulnerabilities are enumerated.
    
    Signed-off-by: Pawan Gupta <pawan.kumar.gupta@linux.intel.com>
    Co-developed-by: Mario Limonciello <mario.limonciello@amd.com>
    Signed-off-by: Mario Limonciello <mario.limonciello@amd.com>
    Signed-off-by: Borislav Petkov (AMD) <bp@alien8.de>
    Acked-by: Dave Hansen <dave.hansen@linux.intel.com>
    Link: https://lore.kernel.org/r/20241025171459.1093-5-mario.limonciello@amd.com

diff --git a/arch/x86/include/asm/topology.h b/arch/x86/include/asm/topology.h
index aef70336d624..9f9376db64e3 100644
--- a/arch/x86/include/asm/topology.h
+++ b/arch/x86/include/asm/topology.h
@@ -114,6 +114,12 @@ enum x86_topology_domains {
 	TOPO_MAX_DOMAIN,
 };
 
+enum x86_topology_cpu_type {
+	TOPO_CPU_TYPE_PERFORMANCE,
+	TOPO_CPU_TYPE_EFFICIENCY,
+	TOPO_CPU_TYPE_UNKNOWN,
+};
+
 struct x86_topology_system {
 	unsigned int	dom_shifts[TOPO_MAX_DOMAIN];
 	unsigned int	dom_size[TOPO_MAX_DOMAIN];
@@ -149,6 +155,9 @@ extern unsigned int __max_threads_per_core;
 extern unsigned int __num_threads_per_package;
 extern unsigned int __num_cores_per_package;
 
+const char *get_topology_cpu_type_name(struct cpuinfo_x86 *c);
+enum x86_topology_cpu_type get_topology_cpu_type(struct cpuinfo_x86 *c);
+
 static inline unsigned int topology_max_packages(void)
 {
 	return __max_logical_packages;